sqlitecluster/SQLiteSequentialNotifier.cpp

```diff
@@ -0,0 +1,74 @@
+#include "SQLiteSequentialNotifier.h"
+
+bool SQLiteSequentialNotifier::waitFor(uint64_t value) {
+    shared_ptr<WaitState> state(nullptr);
```

Comment:
I don't get why this is a `shared_ptr` and not just a value on the stack

Reply:
It's a shared pointer because, to avoid lock contention, we lock the instance mutex `_m` for the whole `SQLiteSequentialNotifier` when we look this up, but we do not hold it once we've found the particular state value that we're interested in.

So we look up `state` for the value we're waiting for (creating it if required), and then we unlock `_m`.

We use a `shared_ptr` instead of a regular pointer because as soon as we unlock `_m` some other thread can call `notifyThrough` on the value we just saved, which with a regular pointer, would delete our data and leave the pointer ready to segfault. With a shared pointer, we still keep the object pointed at until all `shared_ptr`s to it go out of scope, so this object still exists until the end of our function.

Why don't we use a stack variable instead of a `shared_ptr`? Because the thing we're creating a `shared_ptr` contains a condition_variable and mutex that other threads can have locked, or be waiting on. We need to use the same objects that those other threads are using, not a copy created on the stack. In fact, neither `mutex` nor `condition_variable` are copyable, to prevent his problem.

Reply:
The previous comment is not actually an adequate explanation for why we need a shared_ptr here, we could allocate a `WaitState` on the stack and store a pointer to that in the `_valueToPendingThreadMap` map.

We need the `shared_ptr` because this is exactly the case that a reference-counted pointer solves. There are two threads that both need access to the same object, and it's impossible to know which one will finish first, so we need to keep the object until the last thread is finished with it.

Let's assume we create this `WaitState` object here on the stack in `waitFor` and insert a pointer to it in `_valueToPendingThreadMap` (as shown below on line 13, but assume a regular pointer)

We then fall into the `while (true)` loop below that. We are no longer holding `_internalStateMutex` which allows other threads to add/remove `WaitState` objects to the pending list.

In another thread, someone calls `cancel()`. Say this happens immediately before the check of `_globalResult` on line 17 below. This function returns, the local `WaitState` is destroyed, and the other thread calling `cancel` tries to notify our now-destroyed object, and segfaults.

So, the potential "fix" for that is to keep the `WaitState` object in scope as long as `cancel` (or any other call) might need it. Since these changes are guarded by `_internalStateMutex`, before returning, we will need to re-lock `_internalStateMutex`, (which will wait for `cancel` to notify every other thread, because it currently has this mutex locked. It will also serialize each of those threads returning, because they all need to re-lock `_internalStateMutex` in turn) we can then remove our own `WaitState` object from `_valueToPendingThreadMap`, preventing any further calls from accessing that object.

Since `waitFor` is now in charge of removing it's own `WaitState` from the shared list, we remove this from `notifyThrough`. Having it there was nice, because inside a single lock, we could guarantee that we notified each `WaitState` object exactly once, and then removed it from the list of things to notify.

If we put the removal of items from the pending list in `waitFor` instead of `notifyThrough`, then subsequent calls to `notifyThrough` may notify the same objects multiple times, because the thread running `waitFor` may not have had a chance to remove the object from the list yet.

And this just increases contention for `_internalStateMutex` since we now need to synchronize on it twice in `waitFor`.

So, we can use the shared_ptr and just let whichever thread finishes with WaitState last delete it, which is what it's for.